### PR TITLE
feat: add URI parsing utility

### PR DESCRIPTION
**Pull request by aakropotkin**
## Proposed Changes

Adds a slim `parser-util` executable which exposes "flake-ref" URI, Installable URI, and "standard" URI, parsers. Additionally exposes `lockFlake` which includes `originalRef`, `resolvedRef`, and `lockedRef` information used to populate various lockfiles and manifests across the `nix` ecosystem.

Makes this utility accessible from `flox` ( by `PARSER_UTIL_BIN` ) and `flox-bash` ( by `_parser_util` ).

Adds a simple test case which asserts that the executable's `--help` message can be called. This will be replaced with unit tests surrounding the `flox-bash` URI parser helpers and `flox` unit tests surrounding `flox (build|develop|install)`. 

## Current Behavior

Different subsystems each carry their own implementation of an emulated `nix` URI parser. These often cause issues particularly when users attempt to use additional fields on various URIs such as `rev`, `ref`, etc.

## Checks

- [x] All tests pass.
- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] I have accepted the flox [Contributor License Agreement](../blob/main/.github/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](../blob/main/.github/CONTRIBUTORS.csv) file by way of this pull request or one done previously.

## Release Notes

N/A

### CODE REVIEW

**Comment by mkenigs** [comment]
Will flox pull this in directly (i.e. the way you have it with `PARSER_UTIL_BIN`) or via runix?

**Comment by aakropotkin** [comment]
> Will flox(rust) pull this in directly (i.e. the way you have it with `PARSER_UTIL_BIN`) or via runix?

That's not something I know yet, I just wanted to make it available similar to how other executables are. I imagine this would get worked out in the ticket that handles the Rust wrappers

File: parser-util/main.cc [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+  static nlohmann::json
+parseURI( const char * arg )
+{
+  try
+    {
+      nix::ParsedURL       url    = nix::parseURL( arg );
+      nix::ParsedUrlScheme scheme = nix::parseUrlScheme( url.scheme );
+
+      nlohmann::json s = {
```
Comment (zmitchell) [inline parser-util/main.cc:54]: Let's give this a more descriptive name than `s`. I'm not sure what `s` is supposed to stand for here.

File: parser-util/main.cc [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        { "full",        url.scheme       }
+      , { "application", nlohmann::json() }
+      , { "transport",   scheme.transport }
+      };
+      if ( scheme.application.has_value() )
+        {
+          s["application"] = scheme.application.value();
+        }
+
+      nlohmann::json j = {
```
Comment (zmitchell) [inline parser-util/main.cc:64]: Similar to `s` let's give this a more descriptive name.

**Review by zmitchell — COMMENTED** [review]
I've looked over the C++ but it's different enough from Rust that I can't give much insight on that. I'm just trusting that you know what you're doing.

It would be good to have some test cases just as a sanity check to make sure our glue to `nix` works as expected.

File: flox-bash/lib/utils.sh [hunk 3]
```diff
@@ -150,29 +150,31 @@ function hash_commands() {
 # Note that we specifically avoid modifying the PATH environment variable to
 # avoid leaking Nix paths into the commands we invoke.
 # TODO replace each use of $_cut and $_tr with shell equivalents.
-hash_commands \
-	ansifilter awk builtfilter-rs basename bash cat chmod cmp column cp curl cut dasel date dirname \
-	getent gh git grep gum id jq ln man mkdir mktemp mv nix nix-editor nix-store \
-	pwd readlink realpath rm rmdir sed sh sleep sort stat tail tar tee \
-	touch tr uname uuid xargs zgrep semver
+hash_commands                                                                  \
```
[the remark this answers is not in the document]
Comment (aakropotkin) [reply flox-bash/lib/utils.sh:153]: I quoted the executables with `-` characters.
I was running into an issue with them in the test suite so I quoted them to sanity check something and just left it after I finished debugging.

File: parser-util/README.org
[the remark this answers is not in the document]
Comment (aakropotkin) [reply parser-util/README.org:1]: yeah

**Comment by mkenigs** [comment]
> I'm guessing your question is "Are we doing this executable solution or are we going to write CFFI bindings in Rust"? But you might also be asking "do we need to consume this executable in `runix` to write the wrappers there and swap out the underlying implementations"?

Currently in Rust when we need to parse something, we use runix. So regardless of whether we do FFI or JSON, I was wondering if we'll have `parser-util`->`runix`->`flox-rust-sdk`; that way we're not even touching `flox-rust-sdk`, we just swap out the `from_string` implementation in `runix`, which I believe is what @ysndr mentioned

It may impact where we want `parser-util` to live

**Comment by ysndr** [comment]
> Currently in Rust when we need to parse something, we use runix. So regardless of whether we do FFI or JSON, I was wondering if we'll have parser-util->runix->flox-rust-sdk; that way we're not even touching flox-rust-sdk, we just swap out the from_string implementation in runix, which I believe is what @ysndr mentioned

yes that's basically what I meant

Our parsing for flakerefs is in the end provided by runix as `from_str` and derived `Serialize/Deserialze` implementations.

We have some options to integrate the nix parser:

1. dispatch `from_str` and `to_string` to ffi calls (or exec calls to this script)
2. use nix scripts from flox to parse to nix `Attr` or `serde::Value` based types and provide `From<NixFlakeRef> for runix::Flakeref` for when we need the typed variants.
3. mix of both, implement nix parsing as separate methods on runix::FlakeRef, either as `From` implementations or first class methods

(1) would be the least disruptive on the flox side, as for flox the parsing is done transparently
(2) requires only changes in flox, and we dont need to duplicate the c code

**Comment by zmitchell** [comment]
Option 4 is shelling out to this `parser_util` binary and deserializing its output in `from_str`. Alex and I talked yesterday about moving it to a separate repo so that there’s no dependency cycling between the flox/runix repos.On Jul 20, 2023, at 6:05 AM, Yannik Sander ***@***.***> wrote:﻿

Currently in Rust when we need to parse something, we use runix. So regardless of whether we do FFI or JSON, I was wondering if we'll have parser-util->runix->flox-rust-sdk; that way we're not even touching flox-rust-sdk, we just swap out the from_string implementation in runix, which I believe is what @ysndr mentioned

yes that's basically what I meant
Our parsing for flakerefs is in the end provided by runix as from_str and derived Serialize/Deserialze implementations.
We have some options to integrate the nix parser:

dispatch from_str and to_string to ffi calls (or exec calls to this script)
use nix scripts from flox to parse to nix Attr or serde::Value based types and provide From<NixFlakeRef> for runix::Flakeref for when we need the typed variants.
mix of both, implement nix parsing as separate methods on runix::FlakeRef, either as From implementations or first class methods

(1) would be the least disruptive on the flox side, as for flox the parsing is done transparently
(2) requires only changes in flox, and we dont need to duplicate the c code

**Comment by zmitchell** [comment]
Ah, that was covered by option 1. It’s currently 6:20am, my bad.On Jul 20, 2023, at 6:05 AM, Yannik Sander ***@***.***> wrote:﻿

Currently in Rust when we need to parse something, we use runix. So regardless of whether we do FFI or JSON, I was wondering if we'll have parser-util->runix->flox-rust-sdk; that way we're not even touching flox-rust-sdk, we just swap out the from_string implementation in runix, which I believe is what @ysndr mentioned

yes that's basically what I meant
Our parsing for flakerefs is in the end provided by runix as from_str and derived Serialize/Deserialze implementations.
We have some options to integrate the nix parser:

dispatch from_str and to_string to ffi calls (or exec calls to this script)
use nix scripts from flox to parse to nix Attr or serde::Value based types and provide From<NixFlakeRef> for runix::Flakeref for when we need the typed variants.
mix of both, implement nix parsing as separate methods on runix::FlakeRef, either as From implementations or first class methods

(1) would be the least disruptive on the flox side, as for flox the parsing is done transparently
(2) requires only changes in flox, and we dont need to duplicate the c code

File: parser-util/main.cc [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+    }
+  catch( std::exception & e )
+    {
+      std::cerr << e.what() << std::endl;
+      exit( EXIT_FAILURE );
+    }
+
+  /* Unreachable */
+  assert( false );
+  return nlohmann::json();
```
Comment (bryanhonof) [inline parser-util/main.cc:87]: Should we still have code here if you believe it's unreachable?

File: parser-util/main.cc [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
```
Comment (bryanhonof) [inline parser-util/main.cc:125]: Same comment.

File: parser-util/main.cc [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
```
Comment (bryanhonof) [inline parser-util/main.cc:171]: Same comment.

File: parser-util/main.cc [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+  catch( std::exception & e )
+    {
+      // TODO: Catch errors here and make the messages prettier before printing.
+      std::cerr << e.what() << std::endl;
+      exit( EXIT_FAILURE );
+    }
+
+  /* Unreachable */
+  assert( false );
+  return nlohmann::json();
```
Comment (bryanhonof) [inline parser-util/main.cc:235]: Same comment.

**Review by bryanhonof — COMMENTED** [review]
I'm not too familiar with the Nix C++ code, but I don't see any obvious errors right away.

File: parser-util/main.cc (hunk 4 above)
Comment (aakropotkin) [reply parser-util/main.cc:87]: Yeah its to keep the compiler happy. It requires that something be returned in all conditional branches, so even though I believe that block is unreachable I still have to return something.

The assert helps avoid runtime misbehaving if my belief about the block being unreachable was incorrect.

Comment (bryanhonof) [reply parser-util/main.cc:87]: If we're using C++23 we can even do something fancy. 😉
https://en.cppreference.com/w/cpp/utility/unreachable

**Comment by aakropotkin** [comment]
Chatted with the team and opted to move `parser-util` source code to a separate repository where it can be consumed by `flox-bash` directly, and `runix` directly to avoid a cycle between `github:flox/flox` and `github:flox/runix`.

The code was migrated to https://github.com/flox/parser-util.git and references in `github:flox/flox` have been updated to reflect that change.

**Review by zmitchell — APPROVED** [review]

**Pull request merged**